File: LeetCode/Javascript/isPalindromeEg2.js
var isPalindromeEg2 = function(s) {
    s = s.toLowerCase().replace(/[^a-z0-9]/g, "");

    let left = 0;
    let right = s.length - 1;

    while (left < right) {
        if (s[left] !== s[right]) {
            return false;
        }
        left++;
        right--;
    }

    return true;
};

console.log(isPalindromeEg2("A man, a plan, a canal: Panama"))
console.log(isPalindromeEg2('raja'))

//Time: O(n) (each character is processed at most once)
//Space: O(n) (no extra string is created)

File: LeetCode/Go/isPalindromeEg2/isPalindromeEg2.go
package main

import (
	"fmt"
	"regexp"
	"strings"
)

func isPalindromeEg2(s string) bool {
	// Convert to lowercase
	s = strings.ToLower(s)

	// Remove non-alphanumeric characters
	re := regexp.MustCompile(`[^a-z0-9]`)
	s = re.ReplaceAllString(s, "")

	left := 0
	right := len(s) - 1

	for left < right {
		if s[left] != s[right] {
			return false
		}
		left++
		right--
	}

	return true
}

func main() {
	fmt.Println(isPalindromeEg2("A man, a plan, a canal: Panama")) // true
	fmt.Println(isPalindromeEg2("raja"))                           // false
}